extendr-macros/src/wrappers.rs

```diff
@@ -102,17 +100,44 @@ pub fn make_function_wrappers(
     //     }
     // }
     // ```
+    //
     wrappers.push(parse_quote!(
         #[no_mangle]
         #[allow(non_snake_case, clippy::not_unsafe_ptr_arg_deref)]
         pub extern "C" fn #wrap_name(#formal_args) -> extendr_api::SEXP {
-            unsafe {
-                use extendr_api::robj::*;
+            use extendr_api::robj::*;
+            let res_res: std::result::Result<
+                std::result::Result<Robj, extendr_api::Error>,
+                Box<dyn std::any::Any + Send>
+                > = unsafe {
                 #( #convert_args )*
-                extendr_api::handle_panic(#panic_str, ||
-                    extendr_api::Robj::from(#call_name(#actual_args)).get()
-                )
+                std::panic::catch_unwind(||-> std::result::Result<Robj, extendr_api::Error> {
+                    Ok(extendr_api::Robj::from(#call_name(#actual_args)))
+                })
+            };
+            match res_res {
+                Ok(Ok(zz)) => {
+                    return unsafe { zz.get() };
+                }
+                Ok(Err(extendr_err)) => {
+                    let err_string = extendr_err.to_string();
```

Comment:
I'm confused as to how the `Ok(Err)` path occurs. Won't it always be `Err()` or `Ok(Ok())` given this code?

Reply:
`Err(Any)` contains any caught panic. `Ok(Err(Error))` contains any conversion error. Since on new line 330 I replace `unwrap_or_throw_error` with ?-operator to bubble error instead of throwing and panicking. There for the expansion of `#actual_args` has one `?` for each parameter. When a conversion fails  will return from closure on line 114.  Notice the type on line 114 is return type of the closure, not from `catch_unwind()`.

Reply:
if some user function itself returns a Result-type which is Err and since something like #473 is not adopted, `extendr_api::Robj::from` can then also panic in the post-conversion back to R. I define this also as a user induced panic. So a user function will only panic if the user uses panic/unwrap/... or if returning an Err.

Reply:
<img width="616" alt="image" src="https://github.com/extendr/extendr/assets/8015271/f788ad6f-a4af-478b-a39d-1442cfbb6da5">

Reply:
looking at the expansion again, I see can drop the `map_err(|e| ...)` as `?`-operator does that implicitly.
<img width="618" alt="image" src="https://github.com/extendr/extendr/assets/8015271/8c805387-d974-41a4-bf5b-d0086088d5d0">

Reply:
> I'm confused as to how the `Ok(Err)` path occurs. 

Yeah, this is tricky, but this has to be like this because we have to handle `panic!()`.

Just for information, Lionel Henry kindly suggested that I could write some C code to wrap `R_UnwindProtect()` so that no `panic!()` is needed to escape from the cleanup function. I'll experiment this idea and share the result (it probably won't be very soon, so please forget for now).